File: devtools/splashwin.c
#define _GNU_SOURCE
// splashwin — the Discord-updater-splash shape for placement tests: a
// CSD toplevel whose committed buffer is larger than the declared
// window geometry (a shadow margin on all sides), min pinned to max
// in the geometry frame.
// usage: splashwin <w> <h> [margin] [app-id] [late] [parented] [resz] [vismargin]
//   w, h   content size = the declared window geometry, min == max (resz: min only)
//   margin shadow inset: the buffer is (w + 2m) x (h + 2m)
//   late     map first (buffer commit), THEN declare the size limits
//   parented also create a big resizable toplevel first and transient-for it
//   resz     declare the min size only: a resizable CSD window with the same offset
//   vismargin paint the margin as opaque light gray (visible over any background)
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#include <unistd.h>

#include "fixwin-protocol.h"
#include <wayland-client.h>

static struct wl_display   *g_dpy;
static struct wl_compositor *g_comp;
static struct wl_shm       *g_shm;
static struct xdg_wm_base  *g_wm;
static struct wl_surface   *g_surf;
static struct xdg_surface  *g_xsd;
static struct xdg_toplevel *g_top;
static struct wl_buffer    *g_buf;
static struct wl_surface   *g_psurf;
static struct xdg_surface  *g_pxsd;
static struct xdg_toplevel *g_ptop;
static struct wl_buffer    *g_pbuf;
static int                  g_w = 300, g_h = 350, g_m = 10, g_late, g_parented, g_resz, g_vis, g_done;
static const char          *g_id = "splashwin";

static void reg_bind(void *d, struct wl_registry *r, uint32_t id, const char *ifc, uint32_t ver) {
    (void) d;
    (void) ver;
    if (!strcmp(ifc, "wl_compositor") && !g_comp)
        g_comp = wl_registry_bind(r, id, &wl_compositor_interface, 4);
    else if (!strcmp(ifc, "wl_shm") && !g_shm)
        g_shm = wl_registry_bind(r, id, &wl_shm_interface, 1);
    else if (!strcmp(ifc, "xdg_wm_base") && !g_wm)
        g_wm = wl_registry_bind(r, id, &xdg_wm_base_interface, 1);
}
static void reg_remove(void *d, struct wl_registry *r, uint32_t id) {
    (void) d;
    (void) r;
    (void) id;
}
static const struct wl_registry_listener regl = { reg_bind, reg_remove };

static void wm_ping(void *d, struct xdg_wm_base *b, uint32_t s) {
    (void) d;
    xdg_wm_base_pong(b, s);
}
static const struct xdg_wm_base_listener wml = { .ping = wm_ping };

static void top_configure(void *d, struct xdg_toplevel *t, int32_t w, int32_t h, struct wl_array *s) {
    (void) d;
    (void) t;
    (void) w;
    (void) h;
    (void) s;
}
static void top_close(void *d, struct xdg_toplevel *t) {
    (void) d;
    (void) t;
    g_done = 1;
}
static const struct xdg_toplevel_listener topl = { .configure = top_configure, .close = top_close };

int main(int argc, char **argv) {
    g_w  = argc > 1 ? atoi(argv[1]) : 300;
    g_h  = argc > 2 ? atoi(argv[2]) : 350;
    g_m  = argc > 3 ? atoi(argv[3]) : 10;
    if (g_w < 1)
        g_w = 1;
    if (g_h < 1)
        g_h = 1;
    if (g_m < 0)
        g_m = 0;
    if (argc > 4)
        g_id   = argv[4];
    g_late = argc > 5 && !strcmp(argv[5], "late");
    g_parented = argc > 6 && !strcmp(argv[6], "parented");
    g_resz = argc > 7 && !strcmp(argv[7], "resz");
    g_vis  = argc > 8 && !strcmp(argv[8], "vismargin");

    const int BW = g_w + 2 * g_m, BH = g_h + 2 * g_m; // buffer = content + shadow margin

    g_dpy = wl_display_connect(NULL);
    if (!g_dpy)
        return 1;

    struct wl_registry *reg = wl_display_get_registry(g_dpy);
    wl_registry_add_listener(reg, &regl, NULL);
    wl_display_dispatch(g_dpy);
    wl_display_roundtrip(g_dpy);

    if (g_parented) {
        // the main window: big, resizable, same class — covers the center
        g_psurf = wl_compositor_create_surface(g_comp);
        g_pxsd  = xdg_wm_base_get_xdg_surface(g_wm, g_psurf);
        g_ptop  = xdg_surface_get_toplevel(g_pxsd);
        xdg_toplevel_add_listener(g_ptop, &topl, NULL);
        xdg_toplevel_set_app_id(g_ptop, "discord");
        xdg_toplevel_set_title(g_ptop, "Discord");
        const int PW = 1000, PH = 600;
        const size_t PN = (size_t)PW * PH * 4;
        int          pfd = memfd_create("splashwin-parent", 0);
        ftruncate(pfd, PN);
        uint32_t   *ppx = mmap(NULL, PN, PROT_READ | PROT_WRITE, MAP_SHARED, pfd, 0);
        for (size_t i = 0; i < (size_t)PW * PH; i++)
            ppx[i] = 0xff203040;
        struct wl_shm_pool *ppool = wl_shm_create_pool(g_shm, pfd, PN);
        g_pbuf                    = wl_shm_pool_create_buffer(ppool, 0, PW, PH, PW * 4, WL_SHM_FORMAT_ARGB8888);
        wl_shm_pool_destroy(ppool);
        wl_surface_attach(g_psurf, g_pbuf, 0, 0);
        wl_surface_commit(g_psurf);
        wl_display_roundtrip(g_dpy);
        munmap(ppx, PN);
        close(pfd);
    }

    g_surf = wl_compositor_create_surface(g_comp);
    g_xsd  = xdg_wm_base_get_xdg_surface(g_wm, g_surf);
    g_top  = xdg_surface_get_toplevel(g_xsd);
    xdg_wm_base_add_listener(g_wm, &wml, NULL);
    xdg_toplevel_add_listener(g_top, &topl, NULL);
    xdg_toplevel_set_app_id(g_top, g_id);
    xdg_toplevel_set_title(g_top, g_id);
    if (g_parented)
        xdg_toplevel_set_parent(g_top, g_ptop);
    // the CSD content frame: inset from the surface on all sides
    xdg_surface_set_window_geometry(g_xsd, g_m, g_m, g_w, g_h);

    // the fd lives with the process; one fixture window, no reuse
    const size_t N  = (size_t)BW * BH * 4;
    int          fd = memfd_create("splashwin", 0);
    ftruncate(fd, N);
    uint32_t   *px = mmap(NULL, N, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
    for (int y = 0; y < BH; y++) {
        for (int x = 0; x < BW; x++) {
            const int IN = x >= g_m && x < g_m + g_w && y >= g_m && y < g_m + g_h;
            px[y * BW + x] = IN ? 0xff302020 : (g_vis ? 0xffb0b0b0 : 0x00000000);
        }
    }
    struct wl_shm_pool *pool = wl_shm_create_pool(g_shm, fd, N);
    g_buf                    = wl_shm_pool_create_buffer(pool, 0, BW, BH, BW * 4, WL_SHM_FORMAT_ARGB8888);
    wl_shm_pool_destroy(pool);

    if (!g_late) {
        xdg_toplevel_set_min_size(g_top, g_w, g_h);
        if (!g_resz)
            xdg_toplevel_set_max_size(g_top, g_w, g_h); // min == max: the splash shape
    }

    wl_surface_attach(g_surf, g_buf, 0, 0);
    wl_surface_commit(g_surf); // maps (late: without any size limits yet)

    if (g_late) {
        // the race shape: the limits land only after the first map commit
        wl_display_roundtrip(g_dpy);
        xdg_toplevel_set_min_size(g_top, g_w, g_h);
        xdg_toplevel_set_max_size(g_top, g_w, g_h);
        wl_surface_commit(g_surf);
    }

    while (!g_done)
        if (wl_display_dispatch(g_dpy) < 0)
            break;

    xdg_toplevel_destroy(g_top);
    wl_surface_destroy(g_surf);
    wl_buffer_destroy(g_buf);
    if (g_parented) {
        xdg_toplevel_destroy(g_ptop);
        wl_surface_destroy(g_psurf);
        wl_buffer_destroy(g_pbuf);
    }
    xdg_wm_base_destroy(g_wm);
    wl_shm_destroy(g_shm);
    wl_compositor_destroy(g_comp);
    wl_registry_destroy(reg);
    wl_display_disconnect(g_dpy);
    munmap(px, N);
    close(fd);
    return 0;
}

File: devtools/src/bin/splashwin.rs
//! splashwin — the Discord-updater-splash shape for placement tests: a CSD
//! toplevel whose committed buffer is larger than the declared window
//! geometry (a shadow margin on all sides), min pinned to max in the
//! geometry frame.
//!
//! Usage: `splashwin W H [MARGIN] [APP_ID] [late] [parented] [resz]
//! [vismargin] [pinx] [parentonly] [pgeo]`
//! - W, H: content size = the declared window geometry, min == max
//!   (`resz`: min only)
//! - MARGIN: shadow inset; the buffer is (W + 2*M) x (H + 2*M)
//! - `late`: map first (buffer commit), THEN declare the size limits
//! - `parented`: also create a big resizable toplevel first and transient-for it
//! - `resz`: declare the min size only (a resizable CSD window)
//! - `vismargin`: paint the margin opaque light gray (visible over any background)
//! - `pinx`: pin the width only (max = W x 0): the per-axis pin shape
//! - `parentonly`: create only the parent toplevel (no child)
//! - `pgeo`: give the parent an explicit window geometry (0,0,PW,PH)

use std::os::fd::AsFd;

// TEMP use rustix::event::{PollFd, PollFlags, poll};
use rustix::fs::{MemfdFlags, memfd_create};
use rustix::io::write;
use wayland_client::protocol::wl_shm::Format;
use wayland_client::protocol::{
    wl_buffer::WlBuffer, wl_compositor::WlCompositor, wl_registry, wl_shm::WlShm, wl_shm_pool,
};
use wayland_client::{Connection, Dispatch, EventQueue, QueueHandle, delegate_noop};
use wayland_protocols::xdg::shell::client::{xdg_surface, xdg_toplevel, xdg_wm_base};
use xdg_surface::{Event as XdgSurfaceEvent, XdgSurface};
use xdg_toplevel::{Event as ToplevelEvent, XdgToplevel};
use xdg_wm_base::{Event as WmBaseEvent, XdgWmBase};

// the parent's paint color and the child's content/margin colors
const PARENT_COLOR: u32 = 0xff20_3040;
const CONTENT_COLOR: u32 = 0xff30_2020;
const VISIBLE_MARGIN_COLOR: u32 = 0xffb0_b0b0;
const TRANSPARENT: u32 = 0x0000_0000;

fn arg_i64(args: &[String], idx: usize, default: i64) -> i64 {
    args.get(idx)
        .and_then(|a| a.parse().ok())
        .unwrap_or(default)
}

fn flag(args: &[String], idx: usize, name: &str) -> bool {
    args.get(idx).is_some_and(|a| a == name)
}

fn c32(v: i64) -> i32 {
    i32::try_from(v).unwrap_or(i32::MAX)
}

fn fail(code: u8, msg: &str) -> ! {
    eprintln!("{msg}");
    std::process::exit(i32::from(code));
}

// the creation parameters for one fixture window
#[derive(Clone, Copy)]
struct WinSpec<'a> {
    wm: &'a XdgWmBase,
    compositor: &'a WlCompositor,
    shm: &'a WlShm,
    app_id: &'a str,
    width: i64,
    height: i64,
    pixel: &'a dyn Fn(i64, i64) -> u32,
}

// one toplevel: surface + xdg_surface + toplevel + its buffer. The memfd is
// held open for the process lifetime — closing it while the compositor is
// still mapping the buffer destroys the surface.
struct Win {
    surface: Option<wayland_client::protocol::wl_surface::WlSurface>,
    xdg: Option<XdgSurface>,
    toplevel: Option<XdgToplevel>,
    buffer: Option<WlBuffer>,
    _fd: Option<std::os::fd::OwnedFd>,
}

impl Win {
    fn new() -> Self {
        Self {
            surface: None,
            xdg: None,
            toplevel: None,
            buffer: None,
            _fd: None,
        }
    }

    fn create(spec: WinSpec<'_>, qh: &QueueHandle<State>) -> Win {
        let WinSpec {
            wm,
            compositor,
            shm,
            app_id,
            width,
            height,
            pixel,
        } = spec;
        let surface = compositor.create_surface(qh, ());
        let xsd = wm.get_xdg_surface(&surface, qh, ());
        let toplevel = xsd.get_toplevel(qh, ());
        toplevel.set_app_id(app_id.to_string());
        toplevel.set_title(app_id.to_string());

        let byte_count = u64::try_from(width)
            .unwrap_or(u64::MAX)
            .saturating_mul(u64::try_from(height).unwrap_or(u64::MAX))
            .saturating_mul(4);
        let count = usize::try_from(byte_count).unwrap_or(usize::MAX);
        let mut pixels = Vec::with_capacity(count);
        for y in 0..height {
            for x in 0..width {
                pixels.extend_from_slice(&pixel(x, y).to_le_bytes());
            }
        }
        let fd = memfd_create("splashwin", MemfdFlags::CLOEXEC)
            .unwrap_or_else(|err| fail(6, &format!("splashwin: memfd: {err}")));
        let mut offset = 0;
        while offset < pixels.len() {
            let n = write(fd.as_fd(), &pixels[offset..])
                .unwrap_or_else(|err| fail(6, &format!("splashwin: write: {err}")));
            offset += n;
        }
        let size = i32::try_from(pixels.len()).unwrap_or(i32::MAX);
        let pool = shm.create_pool(fd.as_fd(), size, qh, ());
        let buf = pool.create_buffer(
            0,
            i32::try_from(width).unwrap_or(i32::MAX),
            i32::try_from(height).unwrap_or(i32::MAX),
            i32::try_from(width * 4).unwrap_or(i32::MAX),
            Format::Argb8888,
            qh,
            (),
        );
        pool.destroy();
        // attach but do NOT commit: the caller orders the first commit
        // against set_window_geometry / set_parent / size limits, exactly as
        // the C fixture does (the child declares its CSD geometry BEFORE its
        // first frame; the parent maps first and declares geometry after)
        surface.attach(Some(&buf), 0, 0);
        Win {
            surface: Some(surface),
            xdg: Some(xsd),
            toplevel: Some(toplevel),
            buffer: Some(buf),
            _fd: Some(fd),
        }
    }

    fn commit(&self) {
        self.surface.as_ref().unwrap().commit();
    }

    fn teardown(&mut self) {
        if let Some(toplevel) = self.toplevel.take() {
            toplevel.destroy();
        }
        if let Some(xsd) = self.xdg.take() {
            xsd.destroy();
        }
        if let Some(buffer) = self.buffer.take() {
            buffer.destroy();
        }
        if let Some(surface) = self.surface.take() {
            surface.destroy();
        }
    }
}

struct State {
    compositor: Option<WlCompositor>,
    shm: Option<WlShm>,
    wm: Option<XdgWmBase>,
    parent: Win,
    child: Win,
    done: bool,
}

impl Dispatch<wl_registry::WlRegistry, ()> for State {
    fn event(
        state: &mut Self,
        registry: &wl_registry::WlRegistry,
        event: wl_registry::Event,
        _data: &(),
        _conn: &Connection,
        qh: &QueueHandle<Self>,
    ) {
        if let wl_registry::Event::Global {
            name,
            interface,
            version,
        } = event
        {
            match interface.as_str() {
                "wl_compositor" if state.compositor.is_none() => {
                    state.compositor =
                        Some(registry.bind::<WlCompositor, _, _>(name, version.min(4), qh, ()));
                }
                "wl_shm" if state.shm.is_none() => {
                    state.shm = Some(registry.bind::<WlShm, _, _>(name, version.min(1), qh, ()));
                }
                "xdg_wm_base" if state.wm.is_none() => {
                    state.wm = Some(registry.bind::<XdgWmBase, _, _>(name, version.min(1), qh, ()));
                }
                _ => {}
            }
        }
    }
}

impl Dispatch<XdgWmBase, ()> for State {
    fn event(
        _state: &mut Self,
        wm: &XdgWmBase,
        event: WmBaseEvent,
        _data: &(),
        _conn: &Connection,
        _qh: &QueueHandle<Self>,
    ) {
        if let WmBaseEvent::Ping { serial } = event {
            wm.pong(serial);
        }
    }
}

impl Dispatch<XdgSurface, ()> for State {
    fn event(
        _state: &mut Self,
        _xdg: &XdgSurface,
        _event: XdgSurfaceEvent,
        _data: &(),
        _conn: &Connection,
        _qh: &QueueHandle<Self>,
    ) {
    }
}

impl Dispatch<XdgToplevel, ()> for State {
    fn event(
        state: &mut Self,
        _toplevel: &XdgToplevel,
        event: ToplevelEvent,
        _data: &(),
        _conn: &Connection,
        _qh: &QueueHandle<Self>,
    ) {
        if let ToplevelEvent::Close = event {
            state.done = true;
        }
    }
}

delegate_noop!(State: ignore WlCompositor);
delegate_noop!(State: ignore WlShm);
delegate_noop!(State: ignore wl_shm_pool::WlShmPool);
delegate_noop!(State: ignore WlBuffer);
delegate_noop!(State: ignore wayland_client::protocol::wl_surface::WlSurface);

// the CLI shape, parsed once; the flag bools mirror the positional CLI
#[allow(clippy::struct_excessive_bools)]
struct Cli {
    content_w: i64,
    content_h: i64,
    margin: i64,
    app_id: String,
    late: bool,
    parented: bool,
    resz: bool,
    vis: bool,
    pinx: bool,
    parentonly: bool,
    pgeo: bool,
}

fn parse_cli() -> Cli {
    let args: Vec<String> = std::env::args().skip(1).collect();
    Cli {
        content_w: arg_i64(&args, 0, 300).max(1),
        content_h: arg_i64(&args, 1, 350).max(1),
        margin: arg_i64(&args, 2, 10).max(0),
        app_id: args
            .get(3)
            .cloned()
            .unwrap_or_else(|| "splashwin".to_string()),
        late: flag(&args, 4, "late"),
        parented: flag(&args, 5, "parented"),
        resz: flag(&args, 6, "resz"),
        vis: flag(&args, 7, "vismargin"),
        pinx: flag(&args, 8, "pinx"),
        parentonly: flag(&args, 9, "parentonly"),
        pgeo: flag(&args, 10, "pgeo"),
    }
}

fn main() {
    let cli = parse_cli();

    let Ok(conn) = Connection::connect_to_env() else {
        fail(1, "splashwin: no Wayland display");
    };
    let mut queue: EventQueue<State> = conn.new_event_queue();
    let mut state = State {
        compositor: None,
        shm: None,
        wm: None,
        parent: Win::new(),
        child: Win::new(),
        done: false,
    };
    build_windows(&conn, &mut queue, &mut state, &cli);
    run(&conn, &mut queue, &mut state);
}

fn build_windows(conn: &Connection, queue: &mut EventQueue<State>, state: &mut State, cli: &Cli) {
    let qh = queue.handle();
    let content_w = cli.content_w;
    let content_h = cli.content_h;
    let margin = cli.margin;
    let app_id = &cli.app_id;
    let late = cli.late;
    let parented = cli.parented;
    let resz = cli.resz;
    let vis = cli.vis;
    let pinx = cli.pinx;
    let parentonly = cli.parentonly;
    let pgeo = cli.pgeo;
    let _registry = conn.display().get_registry(&qh, ());
    if queue.roundtrip(state).is_err() {
        fail(8, "splashwin: registry roundtrip failed");
    }
    let (compositor, shm, wm) = match (&state.compositor, &state.shm, &state.wm) {
        (Some(c), Some(s), Some(wm)) => (c.clone(), s.clone(), wm.clone()),
        _ => fail(2, "splashwin: compositor/shm/xdg-wm-base unavailable"),
    };

    if parented {
        // the parent: big, resizable, class "discord" — covers the center
        state.parent = Win::create(
            WinSpec {
                wm: &wm,
                compositor: &compositor,
                shm: &shm,
                app_id: "discord",
                width: 1000,
                height: 600,
                pixel: &|_, _| PARENT_COLOR,
            },
            &qh,
        );
        state.parent.commit();
        if queue.roundtrip(state).is_err() {
            fail(9, "splashwin: parent roundtrip failed");
        }
        if pgeo {
            state
                .parent
                .xdg
                .as_ref()
                .unwrap()
                .set_window_geometry(0, 0, 1000, 600);
        }
    }

    if !parentonly {
        let child_pixel = |x: i64, y: i64| -> u32 {
            if x >= margin && x < margin + content_w && y >= margin && y < margin + content_h {
                CONTENT_COLOR
            } else if vis {
                VISIBLE_MARGIN_COLOR
            } else {
                TRANSPARENT
            }
        };
        state.child = Win::create(
            WinSpec {
                wm: &wm,
                compositor: &compositor,
                shm: &shm,
                app_id: app_id.as_str(),
                width: content_w + 2 * margin, // buffer = content + shadow margin
                height: content_h + 2 * margin,
                pixel: &child_pixel,
            },
            &qh,
        );
        if let Some(xsd) = state.child.xdg.as_ref() {
            // the CSD content frame: inset from the surface on all sides
            xsd.set_window_geometry(c32(margin), c32(margin), c32(content_w), c32(content_h));
        }
        if let Some(toplevel) = state.child.toplevel.as_ref() {
            if parented {
                toplevel.set_parent(Some(state.parent.toplevel.as_ref().unwrap()));
            }
            if !late {
                toplevel.set_min_size(c32(content_w), c32(content_h));
                if pinx {
                    toplevel.set_max_size(c32(content_w), 0); // width pinned, height free
                } else if !resz {
                    toplevel.set_max_size(c32(content_w), c32(content_h)); // min == max: the splash shape
                }
            }
        }
        // the child's first frame: AFTER the CSD geometry and size limits
        // (C order: geometry/limits/parent, then attach + commit)
        state.child.commit();
        if late {
            // the race shape: the limits land only after the first map commit
            let toplevel = state
                .child
                .toplevel
                .as_ref()
                .expect("child toplevel")
                .clone();
            if queue.roundtrip(state).is_err() {
                fail(10, "splashwin: late roundtrip failed");
            }
            toplevel.set_min_size(c32(content_w), c32(content_h));
            toplevel.set_max_size(c32(content_w), c32(content_h));
            state.child.surface.as_ref().unwrap().commit();
        }
    }
}

fn run(conn: &Connection, queue: &mut EventQueue<State>, state: &mut State) {
    // blocking dispatch: it flushes the creation burst (binds + pool + buffer
    // + attach + commit) that the setup enqueued after the registry roundtrip.
    // A custom prepare_read/poll loop must flush BETWEEN prepare_read and poll
    // (wl_display_prepare_read_queue does not flush); without it the requests
    // never reach the server and the loop deadlocks waiting for a first frame
    // the compositor can never produce.
    while !state.done {
        if let Err(err) = queue.blocking_dispatch(state) {
            eprintln!("splashwin: dispatch: {err}");
            break;
        }
    }
    state.child.teardown();
    state.parent.teardown();
    if let Some(wm) = state.wm.take() {
        wm.destroy();
    }
    state.shm = None;
    state.compositor = None;
    let _ = conn.flush();
}
